src/pudl_archiver/archivers/msha.py

```diff
@@ -0,0 +1,107 @@
+"""Download MSHA data."""
+
+import logging
+import re
+import typing
+from pathlib import Path
+
+from pudl_archiver.archivers.classes import (
+    AbstractDatasetArchiver,
+    ArchiveAwaitable,
+    ResourceInfo,
+)
+
+logger = logging.getLogger(f"catalystcoop.{__name__}")
+
+BASE_URL = "https://arlweb.msha.gov/OpenGovernmentData/"
+BROWSER_EXT = "OGIMSHA.asp"
+
+URL_107A = "https://arlweb.msha.gov/OpenGovernmentData/107a/"
+EXT_107A = "107aOrders.asp"
+
+datasets = [
+    "Accidents",
+    "AddressofRecord",
+    "AreaSamples",
+    "AssessedViolations",
+    "CivilPenaltyDocketsDecisions",
+    "CoalDustSamples",
+    "Conferences",
+    "ContestedViolations",
+    "ContractorProdQuarterly",
+    "ContractorProdYearly",
+    "ControllerOperatorHistory",
+    "Inspections",
+    "Mines",
+    "MinesProdQuarterly",
+    "MinesProdYearly",
+    "NoiseSamples",
+    "PersonalHealthSamples",
+    "QuartzSamples",
+    "Violations",
+    "107aOrders",
+]
+""" List of expected MSHA tables/definition files."""
+
+
+class MshaArchiver(AbstractDatasetArchiver):
+    """MSHA archiver."""
+
+    name = "mshamines"
+
+    async def get_resources(self) -> ArchiveAwaitable:
+        """Download MSHA resources."""
+        link_pattern = re.compile(
+            r"(?:DataSets\/)([a-zA-Z0-7_]+)(.zip|_Definition_File.txt)"
+        )
+        for link in await self.get_hyperlinks(BASE_URL + BROWSER_EXT, link_pattern):
+            yield self.get_dataset_resource(link, link_pattern.search(link))
+
+        link_pattern = re.compile(r"(?:107a\/)([a-zA-Z0-7]+).xlsx")
+        for link in await self.get_hyperlinks(URL_107A + EXT_107A, link_pattern):
+            yield self.get_107a(link, link_pattern.search(link))
+
+    async def get_dataset_resource(
+        self, link: str, match: typing.Match
+    ) -> tuple[Path, dict]:
+        """Download zip and .txt files."""
+        url = BASE_URL + match.string
+        dataset = match.group(1)
+
+        """Dataset names do not match between data and definition files,
+        so we fix this manually. is tis okay"""
+        dataset = dataset[0].upper() + dataset[1:]
+        dataset = (
+            dataset.replace("_", "")
+            .replace("And", "")
+            .replace("MineS", "Mines")
+            .replace("Annual", "Yearly")
+        )
+        if dataset.endswith("Sample"):
+            dataset = dataset.replace("Sample", "Samples")
```

Comment:
Just to make sure I'm understanding your suggestion - for each 'data' filename, look for the corresponding description file, do some moderate reformatting, and use this to save as the datatype? Rather than conforming files to an existing dictionary?

Reply:
Updated to compare expected urls to a dictionary, raising error if URLs generated from `get_hyperlinks()` are either new or missing from the expected list. Now using snakecase for file names.